packages/atclient/src/atclient.c

```diff
@@ -178,27 +179,41 @@ bool atclient_is_atsign_initialized(const atclient *ctx) {
   /*
    * 1. Validate arguments
    */
-  if(ctx == NULL) {
+  if (ctx == NULL) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "ctx is NULL\n");
     return false;
   }
 
   return ctx->_initialized_fields[ATCLIENT_ATSIGN_INDEX] & ATCLIENT_ATSIGN_INITIALIZED;
 }
 
-int atclient_pkam_authenticate(atclient *ctx, const char *atserver_host, const int atserver_port,
-                               const atclient_atkeys *atkeys, const char *atsign) {
+int atclient_pkam_authenticate(atclient *ctx, const char *atsign, const atclient_atkeys *atkeys,
+                               atclient_pkam_authenticate_options *options) {
 
   int ret = 1; // error by default
+  if(options == NULL){
+    atclient_pkam_authenticate_options options;
+    atclient_pkam_authenticate_options_init(&options);  
+  }
 
   /*
    * 1. Validate arguments
    */
-  if ((ret = atclient_pkam_authenticate_validate_arguments(ctx, atserver_host, atserver_port, atkeys, atsign)) != 0) {
+  if ((ret = atclient_pkam_authenticate_validate_arguments(ctx, atkeys, atsign)) != 0) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_pkam_authenticate_validate_arguments: %d\n", ret);
     return ret;
   }
 
+  /*
+   * 1.1 If atserver_host and atserver_port are null and 0 respectively, then fetch the server host and port.
+   */
+  if (options->atserver_host == NULL && options->atserver_port == 0) {
+    if (atclient_utils_find_atserver_address(options->at_directory_host, options->at_directory_port, atsign,
+                                             &(options->atserver_host), &(options->atserver_port)) != 0) {
+      return ret;
+    }
```

Comment:
If an error happens here, then `atclient_pkam_authenticate_options_free` will never be called.

(This isn't written down in our Contributing spec yet, my apologies), but we have an ongoing convention where for every `*_init` function that is called, `*_free` must be called at the end of the variable's lifetime. This is for future proofing and easy memory management.